compiler/rustc_target/src/abi/call/nvptx64.rs

```diff
@@ -31,3 +45,18 @@ pub fn compute_abi_info<Ty>(fn_abi: &mut FnAbi<'_, Ty>) {
         classify_arg(arg);
     }
 }
+
+pub fn compute_ptx_kernel_abi_info<'a, Ty, C>(cx: &C, fn_abi: &mut FnAbi<'a, Ty>)
+where
+    Ty: TyAbiInterface<'a, C> + Copy,
+    C: HasDataLayout,
+{
+    // Return is not possible for kernels
```

Comment:
Please `bug!` if there are any return types that aren’t `()`, or `!`, I guess?

Reply:
Returning values in kernels doesn't make sense. But right now there are unfortunately no error emitted for attempting to use a return value in a kernel. I've started working on creating a new error which will detect this. 

Am I correct that `bug!` should only be used for compiler bugs? Would it still be beneficial to flag the *user error* by using `bug!`, even before a proper error is implemented? If the answer is yes I will simply add an `is_unit()` and `is_never()` to `rustc_target::abi::TyAbiInterface` and check for those in the place you are referring. I can alternatively add this check in my following PR where the proper error is added as well?

Let me know what you prefer. I'm not familiar enough with rustc to know the right decision.

---

Perhaps unrelated but anyway. My assumption is that the error condition should be detected in `rustc_ast_passes`, do you agree that this is the correct place to implement simple checks on return values?

Reply:
`bug!` is indeed for cases where there's a bug. Your observation that we should be outputting a proper diagnostic here is correct, but ABI code is not the right place for code to do so – typeck would probably be the most appropriate. I'm sure we already have special cases for other ABIs somewhere in there, `rust-call` arguments needing to be a tuple being one IIRC.

Even so, it is a bug if the compiler managed to observe any sort of return value in ABI computation code and so adding a `bug!` is appropriate. Adding a `bug!` now will make the case where there's a `ptx-kernel` function with a return value into an ICE for the time being, but a crash is still better than the compiler doing the wrong thing entirely.

Reply:
Using `bug` (from`rustc_target`) places a dependency on `rustc_middle` (where `bug!` is defined) from `rustc_target`. There is already a dependency on `rustc_target` from `rustc_middle` which makes the new dependency a cyclic dependency. Are there any ways to get around this?

Reply:
`panic!` is a valid alternative.